src/GafferUI/AnimationGadget.cpp

```diff
@@ -0,0 +1,773 @@
+		}
+	}
+
+}
+
+void AnimationGadget::setVisiblePlugs( const std::vector<Gaffer::Plug *> &plugs )
+{
+	m_curvePlugsVisible.clear();
+
+	for( GadgetPtr gadget : m_animationCurves )
+	{
+		CurveGadgetPtr curveGadget = IECore::runTimeCast<CurveGadget>( gadget.get() );
+		if( curveGadget )
+		{
+			removeChild( curveGadget );
+		}
+	}
+	// all these pointers are invalid now anyway
+	m_animationCurves.clear();
+
+	for( const auto &plug : plugs )
+	{
+		if( Gaffer::Animation::CurvePlug *curvePlug = IECore::runTimeCast<Gaffer::Animation::CurvePlug>( plug->getInput()->parent() ) )
+		{
+			std::string name = plug->fullName();  // \todo: clean up name
+			CurveGadget *curveGadget = new CurveGadget( name, curvePlug );
+			addChild( curveGadget );
+			m_animationCurves.push_back( curveGadget );
+
+			m_curvePlugsVisible.push_back( curvePlug );
+		}
+	}
+
+	requestRender();
+}
+
+void AnimationGadget::setEditablePlugs( const std::vector<Gaffer::Plug *> &plugs )
+{
+	m_curvePlugsEditable.clear();
+
+	// \todo: also needs to modify existing selection
+
+	for( auto &plug : plugs )
+	{
+		if( Gaffer::Animation::CurvePlug *curvePlug = IECore::runTimeCast<Gaffer::Animation::CurvePlug>( plug->getInput()->parent() ) )
+		{
+			m_curvePlugsEditable.push_back( curvePlug );
+		}
+	}
+
+	requestRender();
+}
+
+void AnimationGadget::setFrame( float frame )
+{
+	m_currentFrame = frame;
+
+	requestRender();
+}
+
+void AnimationGadget::insertKeyframes()
+{
+	for( auto &curvePlug : m_curvePlugsEditable )
+	{
+		float time = frameToTime( m_currentFrame );
+		if( !curvePlug->hasKey( time ) )
+		{
+			float value = curvePlug->evaluate( time );
+			curvePlug->addKey( Animation::Key( time, value ) );
+		}
+	}
+
+	requestRender();
+}
+
+void AnimationGadget::removeKeyframes()
+{
+	for( const auto &i : m_selectedKeys )
+	{
+		i.second->removeKey( i.first.time );
+	}
+
+	requestRender();
+}
+
+// \todo: This seems a bit messy and it doesn't support Undo.
+// Not sure if we'd ideally hold pointers to the Keys instead to edit them in-place?
+void AnimationGadget::moveKeyframes( const V2f offset )
+{
+
+	std::set<UniqueKey> tmp;
+
+	for( auto &uniqueKey : m_selectedKeys )
+	{
+		// create copy of Key from drag initiation
+		Animation::Key key = m_dragInitialKeys[uniqueKey];
+
+		// remove current key from the curve as it will need to change
+		uniqueKey.second->removeKey( uniqueKey.first.time );
+
+		// apply offsets
+		if( m_moveAxis == MoveAxis::Both || m_moveAxis == MoveAxis::Y )
+		{
+			key.value += offset.y;
+		}
+
+		if( m_moveAxis == MoveAxis::Both || m_moveAxis == MoveAxis::X )
+		{
+			float frame = round( timeToFrame( key.time + offset.x ) );
+			key.time = frameToTime( frame );
+		}
+
+		// add Key back to curve
+		if( uniqueKey.second->hasKey( key.time ) )
+		{
+			// protect existing keys by adding small epsilon
+			// \todo store epsilon somewhere more global
+			key.time += 0.00001;
```

Comment:
Note to myself: This is messing with the snapping to integer frames and potentially causes confusion...

Reply:
Just snap it to the closest frame without a key.